u8timeseries/models/RNN_model.py

```diff
@@ -0,0 +1,469 @@
+
+        :param name: The name of the specific PyTorch RNN module ('RNN', 'GRU' or 'LSTM').
+        :param input_size: The dimensionality of the input time series.
+                           Currently only set to 1 for univariate time series.
+        :param output_length: The number of steps to predict in the future.
+        :param hidden_dim: The number of features in the hidden state h of the RNN module.
+        :param num_layers: The number of recurrent layers.
+        :param num_layers_out_fc: A list containing the hidden dimensions of the layers of the fully connected NN.
+                                  This network connects the last hidden layer of the PyTorch RNN module to the output.
+        :param dropout: The percentage of neurons that are dropped in the non-last RNN layers. Default: 0.
+        """
+
+        super(RNNModule, self).__init__()
+
+        # Defining parameters
+        self.hidden_dim = hidden_dim
+        self.n_layers = num_layers
+        num_layers_out_fc = [] if num_layers_out_fc is None else num_layers_out_fc
+        self.out_len = output_length
+        self.name = name
+
+        # Defining the RNN module
+        self.rnn = getattr(nn, name)(input_size, hidden_dim, num_layers, batch_first=True, dropout=dropout)
+
+        # The RNN module is followed by a fully connected layer, which maps the last hidden layer
+        # to the output of desired length
+        last = hidden_dim
+        feats = []
+        for feature in num_layers_out_fc + [output_length]:
+            feats.append(nn.Linear(last, feature))
+            last = feature
+        self.fc = nn.Sequential(*feats)
+
+    def forward(self, x):
+        # data is of size (batch_size, input_length, input_size)
+        batch_size = x.size(0)
+
+        out, hidden = self.rnn(x)
+
+        # Reshaping the outputs such that it can be fit into the fully connected layer
+        # TODO: we need to test this
+        # if self.many:
+        #     """ Here, we apply the FC network on all the outputs of the network (at all time steps of the sequence)
+        #     """
+        #     predictions = self.fc(out.contiguous().view(-1, self.hidden_dim))
+        #     predictions = predictions.view(batch_size, x.size(1), self.out_len)
+        # else:
+
+        """ Here, we apply the FC network only on the last output point (at the last time step)
+        """
+        if self.name == "LSTM":
+            hidden = hidden[0]
+        predictions = hidden[-1, :, :]
+        predictions = self.fc(predictions)
+        predictions = predictions.view(batch_size, self.out_len, 1)
+
+        # predictions is of size (batch_size, output_length, 1)
+        return predictions
+
+
+class RNNModel(AutoRegressiveModel):
+    def __init__(self,
+                 model: Union[str, nn.Module] = 'RNN',
+                 output_length: int = 1,
+                 input_length: int = 12,
+                 hidden_size: int = 25,
```

Comment:
what if there is more than 1 hidden layer? How should the params look like?

Reply:
The RNNs would be stacked, with same hidden layer sizes. Here, we follow PyTorch, see e.g.: https://pytorch.org/docs/stable/nn.html?highlight=rnn#torch.nn.RNN